Include inactive games in player season box scores

ntsirakis opened the pull request:
Not sure how you wanted to handle inactive games. This is one solution, but you might not like that the dictionary keys are different between active and inactive games. Either way I'd like to help solidify the feature as I am using it for a project of mine.

Edit: I will update tests once the methodology is approved

Signed-off-by: Niko Tsirakis <ntsirakis@github.com>

Commit:
Add flag to include inactive games in player season box scores

Signed-off-by: Niko Tsirakis <ntsirakis@github.com>

jaebradley commented:
@ntsirakis thanks for opening this up

I have two concerns:

1. This is technically a breaking change in behavior - I think adding a flag that allows the addition of inactive box scores as part of the payload where the flag, by default, is off, would lead to less surprising differences in behavior between versions
1. As you point out, the data interface between active and inactive games is different - I think that keeping them identical between the two different states would be best. I think `None` values may be appropriate here.

Out of curiosity, what is your use case for adding this feature?

ntsirakis commented:
> 1. This is technically a breaking change in behavior - I think adding a flag that allows the addition of inactive box scores as part of the payload where the flag, by default, is off, would lead to less surprising differences in behavior between versions

Fair point, but I'm wondering what your reservations are for excluding DNP games by default. To me including them creates a more complete picture, and IMO excluding them is not the behavior I'd expect.

> 2. As you point out, the data interface between active and inactive games is different - I think that keeping them identical between the two different states would be best. I think `None` values may be appropriate here.

Agreed. Would you prefer `None` over 0's?

> Out of curiosity, what is your use case for adding this feature?

I'm working on a fantasy basketball analysis tool. Not only does including DNP's allow me to visualize how reliable a player's health is, it allows me to ask questions like "How does Player A perform when Player B isn't playing?".

jaebradley commented:
> what your reservations are for excluding DNP games by default. To me including them creates a more complete picture, and IMO excluding them is not the behavior I'd expect.

I 💯 % agree - it was a poor decision on my part. My reservations are now mostly that, as I said, it's a breaking change in behavior and for better or worse, my primary goal is to ensure the stability of the existing API and its behavior.


> Agreed. Would you prefer None over 0's?

I would, but I'm open to other thoughts.

ntsirakis commented:
> I agree - it was a poor decision on my part. My reservations are now mostly that, as I said, it's a breaking change in behavior and for better or worse, my primary goal is to ensure the stability of the existing API and its behavior.

I follow. I can add a flag to opt-in in that case and you can figure out if you want to handle that differently in the future.

> I would, but I'm open to other thoughts.

`None` sounds good to me. So, in summary, I'll duplicate all the keys from the standard boxscore, maintain the values for player-agnostic info (opponent, date, etc), drop the counting stats to `None`, and add an `inactive` bool key to both dicts. That sound good?

jaebradley commented:
> add an `inactive` bool key to both dicts.

Not to nitpick but can this be an `active` boolean key? Naming-wise, I'd prefer a positive key name vs. a negative one.

ntsirakis commented:
> Not to nitpick but can this be an `active` boolean key? Naming-wise, I'd prefer a positive key name vs. a negative one.

Ha, funnily enough I held myself back from asking you that exact question so I didn't look too nit-picky. Yes of course.

ntsirakis commented:
Disregard the push, still need to flesh out tests. Can't seem to test them locally for some reason.

ntsirakis commented:
@jaebradley can I ask what your procedure is for running your unittests? Pytest can't seem to find the bbref library. Seeing something similar to the following for every testcase:
```
___________________________________________________________ ERROR collecting tests/unit/parsers/test_team_totals_parser.py ___________________________________________________________
ImportError while importing test module '/home/niko/opensource/basketball_reference_web_scraper/tests/unit/parsers/test_team_totals_parser.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.virtualenvs/pynbafan/lib64/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/unit/parsers/test_team_totals_parser.py:4: in <module>
    from basketball_reference_web_scraper.parsers import TeamTotalsParser, TeamAbbreviationParser
E   ModuleNotFoundError: No module named 'basketball_reference_web_scraper.parsers'
```

I installed pytest-cov in my virtualenv and I'm running `pytest` from within the top-level dir. Tried cd'ing into the second basketball_reference_web_scraper dir as well with the same results. Am I missing something?

Edit: sheesh, fixed it. Apparently having an \_\_init\_\_.py in the tests/ folder breaks testing if you're using pytest. No idea why. Removed that file and it's finally working.

jaebradley commented:
@ntsirakis sorry about the delay in communication - I've actually been feeling under the weather and have just been focussing on getting back to 100%.

Apologies in advance for the test suite - I've perhaps invested too much into integration tests; I'm currently rethinking my testing strategy.

ntsirakis commented:
No rush at all - I’ve been using a patched version of your lib locally so I’m not blocked in any way. Hope you’re feeling better.

jaebradley commented on basketball_reference_web_scraper/client.py:
```diff
@@ -61,12 +61,13 @@ def player_box_scores(day, month, year, output_type=None, output_file_path=None,
 
 def regular_season_player_box_scores(player_identifier, season_end_year, output_type=None, output_file_path=None,
-                                     output_write_option=None, json_options=None):
+                                     output_write_option=None, json_options=None, include_dnp=False):
```
Nit, but mind calling this something like `include_inactive_games`?

A couple reasons why (in my mind) this might be slightly nicer:

1. `dnp` is an acronym, and while most basketball people know what DNP stands for, there might be some that don't
1. DNP seems like it's a pretty specific player designation - for example, Basketball Reference has other specific types of "inactive" player designations

<img width="1075" alt="image" src="https://user-images.githubusercontent.com/8136030/94316818-1224ee00-ff53-11ea-8481-eb6a0096483f.png">

<img width="1050" alt="image" src="https://user-images.githubusercontent.com/8136030/94316903-3f719c00-ff53-11ea-8770-0aa6e3f89a5d.png">

It's unfortunate that `inactive` itself is a player designation - but the term "inactive" maybe encapsulates the concept of a player being "out" or "unavailable to play" or "suspended" more-so than `DNP`?

I'm open to other suggestions.

ntsirakis replied:
Doesn't matter to me, will update to `include_inactive_games`

jaebradley commented on tests/integration/test_player_playoff_box_scores.py:
```diff
@@ -119,3 +120,21 @@ def test_outputting_2019_playoff_box_scores_for_kawhi_leonard_as_csv(self):
+        # 9/8/2020 is not included in his boxscores.
+        results = playoff_player_box_scores(player_identifier="antetgi01", season_end_year=2020)
+        self.assertIsNotNone(results)
+        self.assertEqual(9, len(results))
```
The two other things that might be good to verify (and I can do this after merge if you don't want to) is that the fields have `None` values as well as the CSV and JSON outputs for inactive games look as expected

Same thing for `tests/integration/test_player_regular_season_box_scores.py`.

jaebradley reviewed (commented):
Really close - mostly two things that are discussion / testing-related that I can address post-merge

ntsirakis commented:
I don't remember where we left off with this PR - is this good to merge?

jaebradley commented:
@ntsirakis thanks for being patient.

This should be released as part of `v4.12.0`.

See this `repl.it` example

![image](https://user-images.githubusercontent.com/8136030/103160878-017d5180-47a8-11eb-83ff-ce200346e364.png)

https://repl.it/@jaebradley/Include-Inactive-Games-Parameter#main.py

ntsirakis commented:
@jaebradley no problem at all, thanks for the merge!
